network/p2p/dns/resolver.go

```diff
@@ -109,22 +151,82 @@ func (r *Resolver) LookupTXT(ctx context.Context, txt string) ([]string, error)
 
 // lookupIPAddr encapsulates the logic of resolving a txt through cache.
 func (r *Resolver) lookupTXT(ctx context.Context, txt string) ([]string, error) {
-	if addr, exists, invalidated := r.c.resolveTXTCache(txt); exists {
-		// resolving address from cache
-		r.collector.OnDNSCacheHit()
-		return addr, nil
-	} else if invalidated {
-		r.collector.OnDNSCacheInvalidated()
+	addr, exists, fresh := r.c.resolveTXTCache(txt)
+
+	if !exists {
+		r.collector.OnDNSCacheMiss()
+		return r.lookupResolverForTXTAddr(ctx, txt)
+	}
+
+	if !fresh && r.shouldResolveTXT(txt) {
+		r.unit.Launch(func() {
+			defer r.doneResolvingTXT(txt)
+			_, err := r.lookupResolverForTXTAddr(ctx, txt)
+			if err != nil {
+				// invalidates cached entry when hits error on resolving.
+				invalidated := r.c.invalidateTXTCacheEntry(txt)
+				if invalidated {
+					r.collector.OnDNSCacheInvalidated()
+				}
+			}
+		})
+
 	}
 
-	// resolves txt through underlying resolver
-	r.collector.OnDNSCacheMiss()
+	r.collector.OnDNSCacheHit()
+	return addr, nil
+}
+
+// lookupResolverForIPAddr queries the underlying resolver for the domain and updates the cache if query is successful.
+func (r *Resolver) lookupResolverForTXTAddr(ctx context.Context, txt string) ([]string, error) {
 	addr, err := r.res.LookupTXT(ctx, txt)
 	if err != nil {
 		return nil, err
 	}
 
 	r.c.updateTXTCache(txt, addr) // updates cache
 
-	return addr, err
+	return addr, nil
+}
+
+// shouldResolveIP returns true if there is no other concurrent attempt ongoing for resolving the domain.
+func (r *Resolver) shouldResolveIP(domain string) bool {
+	r.Lock()
+	defer r.Unlock()
+
+	if _, ok := r.processingIPs[domain]; !ok {
+		r.processingIPs[domain] = struct{}{}
+		return true
+	}
+
+	return false
+}
+
```

Comment:
This CAS-like op makes me think this may be a good use for go's [sync.Map](https://www.youtube.com/watch?v=C1EtfDnsdDs), though that's perhaps a place where we want to benchmark rather than speculate.

/cc @jwinkler2083233

Reply:
Francois is right, but since this is not heavily used, the current code is fine as it's written.

Reply:
@huitseeker @jwinkler2083233 implemented `sync.Map` variant on a [different branch](https://github.com/onflow/flow-go/tree/yahya/temp-benchmarking-concurrent-map) and performed benchmarking. Please find the results in the following:

As shown below, at small scale, when thread contention is low (i.e., 300 concurrent operations), the `map` version is around **8 times faster** than `sync.Map`. However, the coin flips swiftly when we scale the operations up:

with 3000 operations, `sync.Map` is **about 2.35 times faster** than `map`.
with 30_000 operations, `sync.Map` is **about 2.66 times faster** than `map`. 
with 300_000 concurrent operations, `sync.Map` is **about 3.27 times faster** than `map`. 

💡  as we scale up on thread contention, `sync.Map` significantly outperforms `map`. It seems like literally, we gonna need `sync.Map` as we scale up the number of nodes. I have it implemented on another branch so moving it to this branch is easy to handle. The question is: 
❓ should we go with `sync.Map` version? 

```
/*
map result
*/

// 300 concurrent operations with map
goos: darwin
goarch: amd64
pkg: github.com/onflow/flow-go/network/p2p/dns
cpu: Intel(R) Core(TM) i5-8279U CPU @ 2.40GHz
BenchmarkCache-8            1000            175202 ns/op

// 3000 concurrent operations with map
goos: darwin
goarch: amd64
pkg: github.com/onflow/flow-go/network/p2p/dns
cpu: Intel(R) Core(TM) i5-8279U CPU @ 2.40GHz
BenchmarkCache-8            1000           1931664 ns/op

// 30_000 concurrent operations with map
goos: darwin
goarch: amd64
pkg: github.com/onflow/flow-go/network/p2p/dns
cpu: Intel(R) Core(TM) i5-8279U CPU @ 2.40GHz
BenchmarkCache-8            1000          24803221 ns/op

// 300_000 concurrent operations with map
goos: darwin
goarch: amd64
pkg: github.com/onflow/flow-go/network/p2p/dns
cpu: Intel(R) Core(TM) i5-8279U CPU @ 2.40GHz
BenchmarkCache-8            1000         318880473 ns/op

/*
sync.Map result.
Averages are taken over 1000 experiments
*/
// 300 concurrent operations with sync.Map
goos: darwin
goarch: amd64
pkg: github.com/onflow/flow-go/network/p2p/dns
cpu: Intel(R) Core(TM) i5-8279U CPU @ 2.40GHz
BenchmarkCache-8            1000             85839 ns/op

// 3000 concurrent operations with sync.Map
goos: darwin
goarch: amd64
pkg: github.com/onflow/flow-go/network/p2p/dns
cpu: Intel(R) Core(TM) i5-8279U CPU @ 2.40GHz
BenchmarkCache-8            1000            812991 ns/op

// 30_000 concurrent operations with sync.Map
goos: darwin
goarch: amd64
pkg: github.com/onflow/flow-go/network/p2p/dns
cpu: Intel(R) Core(TM) i5-8279U CPU @ 2.40GHz
BenchmarkCache-8            1000           9039950 ns/op

// 300_000 concurrent operations with sync.Map
goos: darwin
goarch: amd64
pkg: github.com/onflow/flow-go/network/p2p/dns
cpu: Intel(R) Core(TM) i5-8279U CPU @ 2.40GHz
BenchmarkCache-8            1000          97440807 ns/op
```

Reply:
It's probably not worth buying into `sync.Map` right away, but it would probably be useful to open an issue to keep track of this, and link and persist the branch in which you have experimented with `sync.Map` to it.

Reply:
It's probably not worth buying into `sync.Map` right away, but it would probably be useful to open an issue to keep track of this, and link and persist the branch in which you have experimented with `sync.Map` to it.

This bench is vary nice, thanks for the rigorous experimentation.

Reply:
Issued up here: https://github.com/dapperlabs/flow-go/issues/5855